File: 0001-0500/0496_nextGreaterElement.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        d = {nums2[-1] : -1}
        for i in range(len(nums2)-2, -1, -1):
            if nums2[i] < nums2[i+1]:
                d[nums2[i]] = nums2[i+1]
            else:
                temp = nums2[i+1]
                while temp != -1:
                    if nums2[i] >= temp:
                        temp = d[temp]
                    else:
                        d[nums2[i]] = temp
                        break
                d[nums2[i]] = temp   
        return [d[num] for num in nums1]

# Alternative solution

class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        n = len(nums2)
        stack = []
        next_greater = {}
        i = n - 1
        
        while i >= 0:
            while stack and stack[-1] <= nums2[i]:
                stack.pop()
                
            next_greater[nums2[i]] = stack[-1] if stack else -1
            
            stack.append(nums2[i])
            i -= 1
            
        return [next_greater[i] for i in nums1]

File: 0001-0500/0496_nextGreaterElement.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n1 = nums1.length;
        int n2 = nums2.length;

        HashMap<Integer, Integer> nextGreater = new HashMap<>();
        Stack<Integer> stack = new Stack<>();
        int[] resultSet = new int[n1];

        for (int i = n2 - 1; i >= 0; i--) {
            while (!stack.empty() && stack.peek() <= nums2[i]) {
                stack.pop();
            }

            if (stack.empty()) {
                nextGreater.put(nums2[i], -1);
            } else {
                nextGreater.put(nums2[i], stack.peek());
            }

            stack.push(nums2[i]);
        }

        for (int i = 0; i < n1; i++) {
            resultSet[i] = nextGreater.get(nums1[i]);
        }

        return resultSet;
    }
}
